File: src/tax/us/reconciliation.py
"""
1099-DA Reconciliation Engine.

Matches 1099-DA form entries against calculated disposals to identify
discrepancies and generate actionable guidance for US tax filers.

Matching logic:
  1. Same asset (case-insensitive)
  2. Same date (±1 day tolerance)
  3. Proceeds within 2% tolerance for EXACT, 5% for FUZZY

CRITICAL: All monetary values use decimal.Decimal - NEVER float.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import date, timedelta
from decimal import Decimal, ROUND_HALF_UP
from enum import Enum
from typing import Optional

from src.importers.form_1099da import Form1099DAEntry
from src.calculator.lots import Disposal


# ---------------------------------------------------------------------------
# Enums
# ---------------------------------------------------------------------------

class MatchConfidence(Enum):
    EXACT = "exact"        # date + asset + proceeds all match
    FUZZY = "fuzzy"        # date ±1 day OR proceeds within 5%
    UNMATCHED = "unmatched"


class DiscrepancyType(Enum):
    MISSING_IMPORT = "missing_import"        # on 1099 but not in our data
    DEFI_NOT_ON_1099 = "defi_not_on_1099"    # in our data but not on 1099
    COST_BASIS_DIFF = "cost_basis_diff"
    PROCEEDS_DIFF = "proceeds_diff"


# ---------------------------------------------------------------------------
# Data models
# ---------------------------------------------------------------------------

@dataclass
class Discrepancy:
    """A single discrepancy between 1099-DA and our calculated data."""
    type: DiscrepancyType
    description: str
    form_value: Optional[Decimal]
    our_value: Optional[Decimal]
    guidance: str  # "What to do" text


@dataclass
class MatchedEntry:
    """A matched pair: 1099-DA entry ↔ our disposal."""
    form_entry: Form1099DAEntry
    our_disposal: Disposal
    confidence: MatchConfidence
    discrepancies: list[Discrepancy] = field(default_factory=list)


@dataclass
class ReconciliationResult:
    """Complete reconciliation output."""
    matched: list[MatchedEntry]
    unmatched_1099: list[Form1099DAEntry]
    unmatched_ours: list[Disposal]
    discrepancies: list[Discrepancy]
    total_1099_proceeds: Decimal
    total_our_proceeds: Decimal
    match_rate: Decimal  # percentage matched (0-100)
    summary_text: str


# ---------------------------------------------------------------------------
# Matching helpers
# ---------------------------------------------------------------------------

def _normalize_asset(asset: str) -> str:
    """Normalize asset symbol for comparison."""
    return asset.strip().upper()


def _dates_match_exact(d1: date, d2: date) -> bool:
    """Check if two dates are the same."""
    return d1 == d2


def _dates_match_fuzzy(d1: date, d2: date, tolerance_days: int = 1) -> bool:
    """Check if two dates are within tolerance."""
    return abs((d1 - d2).days) <= tolerance_days


def _proceeds_match_exact(p1: Decimal, p2: Decimal, tolerance_pct: Decimal = Decimal("0.02")) -> bool:
    """Check if proceeds match within tolerance percentage (default 2%)."""
    if p1 == Decimal("0") and p2 == Decimal("0"):
        return True
    if p1 == Decimal("0") or p2 == Decimal("0"):
        return False
    diff_pct = abs(p1 - p2) / max(abs(p1), abs(p2))
    return diff_pct <= tolerance_pct


def _proceeds_match_fuzzy(p1: Decimal, p2: Decimal, tolerance_pct: Decimal = Decimal("0.05")) -> bool:
    """Check if proceeds match within fuzzy tolerance (default 5%)."""
    return _proceeds_match_exact(p1, p2, tolerance_pct)


def _disposal_date(disposal: Disposal) -> date:
    """Extract date from a Disposal (which stores datetime)."""
    if hasattr(disposal.date, "date"):
        return disposal.date.date()
    return disposal.date


# ---------------------------------------------------------------------------
# Guidance text generators
# ---------------------------------------------------------------------------

def _guidance_missing_import(broker: str) -> str:
    return (
        f"This transaction appears on your 1099-DA from {broker} but not in "
        f"your imported data. Import your {broker} transactions to resolve."
    )


def _guidance_defi_not_on_1099() -> str:
    return (
        "This DeFi transaction is not reported on any 1099-DA. "
        "You must self-report it on Form 8949."
    )


def _guidance_cost_basis_diff(form_value: Decimal, our_value: Decimal) -> str:
    return (
        f"Your 1099-DA shows cost basis of ${form_value:,.2f} but we calculated "
        f"${our_value:,.2f}. This may be due to transfers between wallets "
        f"affecting lot ordering."
    )


def _guidance_proceeds_diff(diff: Decimal) -> str:
    return (
        f"Proceeds differ by ${abs(diff):,.2f}. This is typically due to "
        f"rounding or fee treatment differences."
    )


# ---------------------------------------------------------------------------
# Core reconciliation engine
# ---------------------------------------------------------------------------

def reconcile(
    form_entries: list[Form1099DAEntry],
    our_disposals: list[Disposal],
) -> ReconciliationResult:
    """
    Reconcile 1099-DA form entries against our calculated disposals.

    Matching priority:
      1. EXACT: same asset + same date + proceeds within 2%
      2. FUZZY: same asset + date ±1 day + proceeds within 5%
      3. UNMATCHED: no match found

    Returns a ReconciliationResult with matched pairs, unmatched entries
    on both sides, discrepancies, and summary statistics.
    """
    matched: list[MatchedEntry] = []
    all_discrepancies: list[Discrepancy] = []

    # Track which disposals have been claimed
    available_disposals = list(our_disposals)

    # Track unmatched 1099 entries
    unmatched_1099: list[Form1099DAEntry] = []

    for form_entry in form_entries:
        best_match: Optional[Disposal] = None
        best_confidence = MatchConfidence.UNMATCHED
        best_index = -1

        form_asset = _normalize_asset(form_entry.asset)
        form_date = form_entry.date_sold
        form_proceeds = form_entry.proceeds

        for idx, disposal in enumerate(available_disposals):
            disp_asset = _normalize_asset(disposal.token)
            disp_date = _disposal_date(disposal)
            disp_proceeds = disposal.proceeds_usd

            # Must be same asset
            if form_asset != disp_asset:
                continue

            # Check EXACT match first
            if (_dates_match_exact(form_date, disp_date) and
                    _proceeds_match_exact(form_proceeds, disp_proceeds)):
                best_match = disposal
                best_confidence = MatchConfidence.EXACT
                best_index = idx
                break  # exact is best possible

            # Check FUZZY match
            if (_dates_match_fuzzy(form_date, disp_date) and
                    _proceeds_match_fuzzy(form_proceeds, disp_proceeds)):
                if best_confidence != MatchConfidence.EXACT:
                    best_match = disposal
                    best_confidence = MatchConfidence.FUZZY
                    best_index = idx

        if best_match is not None and best_index >= 0:
            # Remove from available pool
            available_disposals.pop(best_index)

            # Check for discrepancies within matched pair
            entry_discrepancies: list[Discrepancy] = []

            # Check cost basis difference
            if form_entry.cost_basis is not None and best_match.cost_basis_usd is not None:
                if not _proceeds_match_exact(form_entry.cost_basis, best_match.cost_basis_usd):
                    disc = Discrepancy(
                        type=DiscrepancyType.COST_BASIS_DIFF,
                        description=(
                            f"Cost basis mismatch for {form_asset}: "
                            f"1099-DA=${form_entry.cost_basis:,.2f}, "
                            f"ours=${best_match.cost_basis_usd:,.2f}"
                        ),
                        form_value=form_entry.cost_basis,
                        our_value=best_match.cost_basis_usd,
                        guidance=_guidance_cost_basis_diff(
                            form_entry.cost_basis, best_match.cost_basis_usd
                        ),
                    )
                    entry_discrepancies.append(disc)
                    all_discrepancies.append(disc)

            # Check proceeds difference (even in fuzzy matches)
            proceeds_diff = form_entry.proceeds - best_match.proceeds_usd
            if abs(proceeds_diff) > Decimal("0.01"):
                disc = Discrepancy(
                    type=DiscrepancyType.PROCEEDS_DIFF,
                    description=(
                        f"Proceeds differ for {form_asset}: "
                        f"1099-DA=${form_entry.proceeds:,.2f}, "
                        f"ours=${best_match.proceeds_usd:,.2f}"
                    ),
                    form_value=form_entry.proceeds,
                    our_value=best_match.proceeds_usd,
                    guidance=_guidance_proceeds_diff(proceeds_diff),
                )
                entry_discrepancies.append(disc)
                all_discrepancies.append(disc)

            matched.append(MatchedEntry(
                form_entry=form_entry,
                our_disposal=best_match,
                confidence=best_confidence,
                discrepancies=entry_discrepancies,
            ))
        else:
            # No match found - missing import
            unmatched_1099.append(form_entry)
            disc = Discrepancy(
                type=DiscrepancyType.MISSING_IMPORT,
                description=(
                    f"{form_asset} sold on {form_entry.date_sold} for "
                    f"${form_entry.proceeds:,.2f} appears on 1099-DA from "
                    f"{form_entry.broker_name} but not in your imported data."
                ),
                form_value=form_entry.proceeds,
                our_value=None,
                guidance=_guidance_missing_import(form_entry.broker_name),
            )
            all_discrepancies.append(disc)

    # Remaining disposals = DeFi/self-custody not on any 1099
    unmatched_ours: list[Disposal] = available_disposals
    for disposal in unmatched_ours:
        disc = Discrepancy(
            type=DiscrepancyType.DEFI_NOT_ON_1099,
            description=(
                f"{_normalize_asset(disposal.token)} disposed on "
                f"{_disposal_date(disposal)} for ${disposal.proceeds_usd:,.2f} "
                f"is not reported on any 1099-DA."
            ),
            form_value=None,
            our_value=disposal.proceeds_usd,
            guidance=_guidance_defi_not_on_1099(),
        )
        all_discrepancies.append(disc)

    # Compute summary statistics
    total_1099 = sum((e.proceeds for e in form_entries), Decimal("0"))
    total_ours = sum((d.proceeds_usd for d in our_disposals), Decimal("0"))

    total_entries = len(form_entries)
    match_rate = (
        (Decimal(len(matched)) / Decimal(total_entries) * Decimal("100"))
        if total_entries > 0
        else Decimal("100")
    ).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)

    # Build summary text
    summary_lines = [
        f"Reconciliation Summary",
        f"=" * 40,
        f"Total 1099-DA entries:     {len(form_entries)}",
        f"Total calculated disposals: {len(our_disposals)}",
        f"",
        f"Matched (exact):           {sum(1 for m in matched if m.confidence == MatchConfidence.EXACT)}",
        f"Matched (fuzzy):           {sum(1 for m in matched if m.confidence == MatchConfidence.FUZZY)}",
        f"Unmatched on 1099-DA:      {len(unmatched_1099)}",
        f"Unmatched in our data:     {len(unmatched_ours)}",
        f"",
        f"Match rate:                {match_rate}%",
        f"",
        f"Total 1099-DA proceeds:    ${total_1099:,.2f}",
        f"Total our proceeds:        ${total_ours:,.2f}",
        f"Difference:                ${abs(total_1099 - total_ours):,.2f}",
        f"",
        f"Discrepancies found:       {len(all_discrepancies)}",
    ]

    if unmatched_1099:
        summary_lines.append("")
        summary_lines.append("ACTION NEEDED: Import missing transactions")
        brokers = set(e.broker_name for e in unmatched_1099)
        for broker in sorted(brokers):
            count = sum(1 for e in unmatched_1099 if e.broker_name == broker)
            summary_lines.append(f"  - {broker}: {count} unmatched transaction(s)")

    if unmatched_ours:
        summary_lines.append("")
        summary_lines.append("ACTION NEEDED: Self-report DeFi disposals on Form 8949")
        summary_lines.append(
            f"  - {len(unmatched_ours)} disposal(s) not on any 1099-DA"
        )

    return ReconciliationResult(
        matched=matched,
        unmatched_1099=unmatched_1099,
        unmatched_ours=unmatched_ours,
        discrepancies=all_discrepancies,
        total_1099_proceeds=total_1099,
        total_our_proceeds=total_ours,
        match_rate=match_rate,
        summary_text="\n".join(summary_lines),
    )

File: frontend/src/engine/tax/us/reconciliation.ts
/**
 * 1099-DA Reconciliation Engine.
 *
 * Ported from src/tax/us/reconciliation.py.
 *
 * Matches 1099-DA form entries against calculated disposals to identify
 * discrepancies and generate actionable guidance for US tax filers.
 *
 * Matching logic:
 *   1. Same asset (case-insensitive)
 *   2. Same date (+/-1 day tolerance)
 *   3. Proceeds within 2% tolerance for EXACT, 5% for FUZZY
 *
 * CRITICAL: All monetary values use decimal.js Decimal -- NEVER native number for money.
 */

import Decimal from 'decimal.js';
import type { Disposal } from '../../types';

// ---------------------------------------------------------------------------
// Enums
// ---------------------------------------------------------------------------

export enum MatchConfidence {
  EXACT = 'exact',
  FUZZY = 'fuzzy',
  UNMATCHED = 'unmatched',
}

export enum DiscrepancyType {
  MISSING_IMPORT = 'missing_import',
  DEFI_NOT_ON_1099 = 'defi_not_on_1099',
  COST_BASIS_DIFF = 'cost_basis_diff',
  PROCEEDS_DIFF = 'proceeds_diff',
}

// ---------------------------------------------------------------------------
// Data models
// ---------------------------------------------------------------------------

/** A single line item from a 1099-DA form. */
export interface Form1099DAEntry {
  asset: string;
  dateAcquired: string | null; // ISO date or null for 'VARIOUS'
  dateSold: string;            // ISO date
  proceeds: Decimal;
  costBasis: Decimal | null;
  gainLoss: Decimal | null;
  brokerName: string;
  isCovered: boolean;
  rawRow: Record<string, string>;
}

/** A single discrepancy between 1099-DA and our calculated data. */
export interface Discrepancy {
  type: DiscrepancyType;
  description: string;
  formValue: Decimal | null;
  ourValue: Decimal | null;
  guidance: string;
}

/** A matched pair: 1099-DA entry <-> our disposal. */
export interface MatchedEntry {
  formEntry: Form1099DAEntry;
  ourDisposal: Disposal;
  confidence: MatchConfidence;
  discrepancies: Discrepancy[];
}

/** Complete reconciliation output. */
export interface ReconciliationResult {
  matched: MatchedEntry[];
  unmatched1099: Form1099DAEntry[];
  unmatchedOurs: Disposal[];
  discrepancies: Discrepancy[];
  total1099Proceeds: Decimal;
  totalOurProceeds: Decimal;
  matchRate: Decimal;
  summaryText: string;
}

// ---------------------------------------------------------------------------
// Matching helpers
// ---------------------------------------------------------------------------

function normalizeAsset(asset: string): string {
  return asset.trim().toUpperCase();
}

function parseDateOnly(iso: string): Date {
  return new Date(iso.slice(0, 10) + 'T00:00:00Z');
}

function datesMatchExact(d1: Date, d2: Date): boolean {
  return d1.getTime() === d2.getTime();
}

function datesMatchFuzzy(d1: Date, d2: Date, toleranceDays: number = 1): boolean {
  const diffMs = Math.abs(d1.getTime() - d2.getTime());
  return diffMs <= toleranceDays * 86400000;
}

function proceedsMatchExact(
  p1: Decimal, p2: Decimal, tolerancePct: Decimal = new Decimal('0.02'),
): boolean {
  if (p1.isZero() && p2.isZero()) return true;
  if (p1.isZero() || p2.isZero()) return false;
  const diffPct = p1.minus(p2).abs().div(Decimal.max(p1.abs(), p2.abs()));
  return diffPct.lte(tolerancePct);
}

function proceedsMatchFuzzy(
  p1: Decimal, p2: Decimal, tolerancePct: Decimal = new Decimal('0.05'),
): boolean {
  return proceedsMatchExact(p1, p2, tolerancePct);
}

// ---------------------------------------------------------------------------
// Guidance text generators
// ---------------------------------------------------------------------------

function guidanceMissingImport(broker: string): string {
  return (
    `This transaction appears on your 1099-DA from ${broker} but not in ` +
    `your imported data. Import your ${broker} transactions to resolve.`
  );
}

function guidanceDefiNotOn1099(): string {
  return (
    'This DeFi transaction is not reported on any 1099-DA. ' +
    'You must self-report it on Form 8949.'
  );
}

function guidanceCostBasisDiff(formValue: Decimal, ourValue: Decimal): string {
  return (
    `Your 1099-DA shows cost basis of $${formValue.toFixed(2)} but we calculated ` +
    `$${ourValue.toFixed(2)}. This may be due to transfers between wallets ` +
    `affecting lot ordering.`
  );
}

function guidanceProceedsDiff(diff: Decimal): string {
  return (
    `Proceeds differ by $${diff.abs().toFixed(2)}. This is typically due to ` +
    `rounding or fee treatment differences.`
  );
}

// ---------------------------------------------------------------------------
// Core reconciliation engine
// ---------------------------------------------------------------------------

/**
 * Reconcile 1099-DA form entries against our calculated disposals.
 *
 * Matching priority:
 *   1. EXACT: same asset + same date + proceeds within 2%
 *   2. FUZZY: same asset + date +/-1 day + proceeds within 5%
 *   3. UNMATCHED: no match found
 *
 * @param formEntries   - Array of Form1099DAEntry objects
 * @param ourDisposals  - Array of Disposal objects
 * @returns ReconciliationResult
 */
export function reconcile(
  formEntries: Form1099DAEntry[],
  ourDisposals: Disposal[],
): ReconciliationResult {
  const matched: MatchedEntry[] = [];
  const allDiscrepancies: Discrepancy[] = [];

  // Track available disposals
  const availableDisposals = [...ourDisposals];
  const unmatched1099: Form1099DAEntry[] = [];

  for (const formEntry of formEntries) {
    let bestMatch: Disposal | null = null;
    let bestConfidence = MatchConfidence.UNMATCHED;
    let bestIndex = -1;

    const formAsset = normalizeAsset(formEntry.asset);
    const formDate = parseDateOnly(formEntry.dateSold);
    const formProceeds = formEntry.proceeds;

    for (let idx = 0; idx < availableDisposals.length; idx++) {
      const disposal = availableDisposals[idx];
      const dispAsset = normalizeAsset(disposal.token);
      const dispDate = parseDateOnly(disposal.date);
      const dispProceeds = disposal.proceedsUsd;

      // Must be same asset
      if (formAsset !== dispAsset) continue;

      // Check EXACT match first
      if (datesMatchExact(formDate, dispDate) &&
          proceedsMatchExact(formProceeds, dispProceeds)) {
        bestMatch = disposal;
        bestConfidence = MatchConfidence.EXACT;
        bestIndex = idx;
        break; // exact is best possible
      }

      // Check FUZZY match (only upgrade if we haven't found an exact or fuzzy yet)
      if (datesMatchFuzzy(formDate, dispDate) &&
          proceedsMatchFuzzy(formProceeds, dispProceeds)) {
        if (bestConfidence === MatchConfidence.UNMATCHED) {
          bestMatch = disposal;
          bestConfidence = MatchConfidence.FUZZY;
          bestIndex = idx;
        }
      }
    }

    if (bestMatch !== null && bestIndex >= 0) {
      availableDisposals.splice(bestIndex, 1);

      const entryDiscrepancies: Discrepancy[] = [];

      // Check cost basis difference
      if (formEntry.costBasis !== null) {
        if (!proceedsMatchExact(formEntry.costBasis, bestMatch.costBasisUsd)) {
          const disc: Discrepancy = {
            type: DiscrepancyType.COST_BASIS_DIFF,
            description:
              `Cost basis mismatch for ${formAsset}: ` +
              `1099-DA=$${formEntry.costBasis.toFixed(2)}, ` +
              `ours=$${bestMatch.costBasisUsd.toFixed(2)}`,
            formValue: formEntry.costBasis,
            ourValue: bestMatch.costBasisUsd,
            guidance: guidanceCostBasisDiff(formEntry.costBasis, bestMatch.costBasisUsd),
          };
          entryDiscrepancies.push(disc);
          allDiscrepancies.push(disc);
        }
      }

      // Check proceeds difference
      const proceedsDiff = formEntry.proceeds.minus(bestMatch.proceedsUsd);
      if (proceedsDiff.abs().gt('0.01')) {
        const disc: Discrepancy = {
          type: DiscrepancyType.PROCEEDS_DIFF,
          description:
            `Proceeds differ for ${formAsset}: ` +
            `1099-DA=$${formEntry.proceeds.toFixed(2)}, ` +
            `ours=$${bestMatch.proceedsUsd.toFixed(2)}`,
          formValue: formEntry.proceeds,
          ourValue: bestMatch.proceedsUsd,
          guidance: guidanceProceedsDiff(proceedsDiff),
        };
        entryDiscrepancies.push(disc);
        allDiscrepancies.push(disc);
      }

      matched.push({
        formEntry,
        ourDisposal: bestMatch,
        confidence: bestConfidence,
        discrepancies: entryDiscrepancies,
      });
    } else {
      unmatched1099.push(formEntry);
      const disc: Discrepancy = {
        type: DiscrepancyType.MISSING_IMPORT,
        description:
          `${formAsset} sold on ${formEntry.dateSold} for ` +
          `$${formEntry.proceeds.toFixed(2)} appears on 1099-DA from ` +
          `${formEntry.brokerName} but not in your imported data.`,
        formValue: formEntry.proceeds,
        ourValue: null,
        guidance: guidanceMissingImport(formEntry.brokerName),
      };
      allDiscrepancies.push(disc);
    }
  }

  // Remaining disposals = DeFi/self-custody not on any 1099
  const unmatchedOurs = availableDisposals;
  for (const disposal of unmatchedOurs) {
    const disc: Discrepancy = {
      type: DiscrepancyType.DEFI_NOT_ON_1099,
      description:
        `${normalizeAsset(disposal.token)} disposed on ` +
        `${disposal.date.slice(0, 10)} for $${disposal.proceedsUsd.toFixed(2)} ` +
        `is not reported on any 1099-DA.`,
      formValue: null,
      ourValue: disposal.proceedsUsd,
      guidance: guidanceDefiNotOn1099(),
    };
    allDiscrepancies.push(disc);
  }

  // Compute summary statistics
  let total1099 = new Decimal('0');
  for (const e of formEntries) total1099 = total1099.plus(e.proceeds);

  let totalOurs = new Decimal('0');
  for (const d of ourDisposals) totalOurs = totalOurs.plus(d.proceedsUsd);

  const totalEntries = formEntries.length;
  const matchRate = totalEntries > 0
    ? new Decimal(matched.length).div(totalEntries).times(100).toDecimalPlaces(2)
    : new Decimal('100');

  // Build summary text
  const summaryLines: string[] = [
    'Reconciliation Summary',
    '='.repeat(40),
    `Total 1099-DA entries:     ${formEntries.length}`,
    `Total calculated disposals: ${ourDisposals.length}`,
    '',
    `Matched (exact):           ${matched.filter((m) => m.confidence === MatchConfidence.EXACT).length}`,
    `Matched (fuzzy):           ${matched.filter((m) => m.confidence === MatchConfidence.FUZZY).length}`,
    `Unmatched on 1099-DA:      ${unmatched1099.length}`,
    `Unmatched in our data:     ${unmatchedOurs.length}`,
    '',
    `Match rate:                ${matchRate.toString()}%`,
    '',
    `Total 1099-DA proceeds:    $${total1099.toFixed(2)}`,
    `Total our proceeds:        $${totalOurs.toFixed(2)}`,
    `Difference:                $${total1099.minus(totalOurs).abs().toFixed(2)}`,
    '',
    `Discrepancies found:       ${allDiscrepancies.length}`,
  ];

  if (unmatched1099.length > 0) {
    summaryLines.push('');
    summaryLines.push('ACTION NEEDED: Import missing transactions');
    const brokers = Array.from(new Set(unmatched1099.map((e) => e.brokerName)));
    for (const broker of brokers.sort()) {
      const count = unmatched1099.filter((e) => e.brokerName === broker).length;
      summaryLines.push(`  - ${broker}: ${count} unmatched transaction(s)`);
    }
  }

  if (unmatchedOurs.length > 0) {
    summaryLines.push('');
    summaryLines.push('ACTION NEEDED: Self-report DeFi disposals on Form 8949');
    summaryLines.push(`  - ${unmatchedOurs.length} disposal(s) not on any 1099-DA`);
  }

  return {
    matched,
    unmatched1099: unmatched1099,
    unmatchedOurs: unmatchedOurs,
    discrepancies: allDiscrepancies,
    total1099Proceeds: total1099,
    totalOurProceeds: totalOurs,
    matchRate,
    summaryText: summaryLines.join('\n'),
  };
}
